pkg/tsdb/prometheus/prometheus.go

```diff
@@ -292,17 +305,25 @@ func (s *Service) parseQuery(queryContext *backend.QueryDataRequest, dsInfo *Dat
 		expr = strings.ReplaceAll(expr, "$__rate_interval", intervalv2.FormatDuration(calculateRateInterval(interval, dsInfo.TimeInterval, s.intervalCalculator)))
 
 		queryType := Range
+		if model.InstantQuery {
+			queryType = Instant
+		}
 
-		// Align query range to step. It rounds start and end down to a multiple of step.
-		start := int64(math.Floor((float64(query.TimeRange.From.Unix()+model.OffsetSec)/interval.Seconds()))*interval.Seconds() - float64(model.OffsetSec))
-		end := int64(math.Floor((float64(query.TimeRange.To.Unix()+model.OffsetSec)/interval.Seconds()))*interval.Seconds() - float64(model.OffsetSec))
+		start := query.TimeRange.From
+		end := query.TimeRange.To
+
+		if queryType == Range {
+			// Align query range to step. It rounds start and end down to a multiple of step.
+			start = time.Unix(int64(math.Floor((float64(query.TimeRange.From.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
+			end = time.Unix(int64(math.Floor((float64(query.TimeRange.To.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
+		}
```

Comment:
what about something like:

```go
		var queryType PrometheusQueryType
		var start time.Time
		var end time.Time

		if(model.RangeQuery != model.InstantQuery) {
			return nil, fmt.Errorf("the provided query is not valid. expected one of `range` and `instant` to be true.")
		}

		if model.RangeQuery {
			queryType = Range
			// Align query range to step. It rounds start and end down to a multiple of step.
			start = time.Unix(int64(math.Floor((float64(query.TimeRange.From.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
			end = time.Unix(int64(math.Floor((float64(query.TimeRange.To.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
		} else if model.InstantQuery {
			queryType = Instant
			start = query.TimeRange.From
			end = query.TimeRange.To	
		}
```

that way, since we don't support running both instant and range queries in the same request, we also apply validation on BE.
and *maybe* (opinionanted) it's a bit easier to read

WDYT?

Reply:
to expand a bit more, the reason behind this is that this whole backend implementation can be used as an API (https://grafana.com/docs/grafana/latest/http_api/data_source/#query-a-data-source-by-id) and even though we now grafana FE itself will only use it the proper way it might be worth it to be on the safe side

Reply:
got it! so I assume this would be safe as well for alerting, but provide 
a degree of validation when trying to perform both a range and instant query in the same request?

```go

if model.RangeQuery && model.InstantQuery {
  // not sure about the message here
  return nil, fmt.Errorf("the provided query is not valid. expected only one of `range` and `instant` to be true.")
}

var queryType PrometheusQueryType
var start time.Time
var end time.Time
     
if model.InstantQuery {
  queryType = Instant
 start = query.TimeRange.From
  end = query.TimeRange.To	
} else {
  queryType = Range
  // Align query range to step. It rounds start and end down to a multiple of step.
  start = time.Unix(int64(math.Floor((float64(query.TimeRange.From.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
  end = time.Unix(int64(math.Floor((float64(query.TimeRange.To.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
}

Reply:
yes, this could be used. But as mentioned this morning,  I would like to discuss other thought I have, whenever you have time. 🙂